cpp/src/arrow/util/iterator_test.cc

```diff
@@ -940,4 +942,144 @@ TEST(TestAsyncIteratorTransform, SkipSome) {
   AssertAsyncGeneratorMatch({"1", "3"}, std::move(filtered));
 }
 
+TEST(PushGenerator, Empty) {
+  PushGenerator<TestInt> gen;
+  auto producer = gen.producer();
+
+  auto fut = gen();
+  AssertNotFinished(fut);
+  producer.Close();
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), fut);
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
+
+  // Close idempotent
+  fut = gen();
+  producer.Close();
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), fut);
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
+}
+
+TEST(PushGenerator, Success) {
+  PushGenerator<TestInt> gen;
+  auto producer = gen.producer();
+  std::vector<Future<TestInt>> futures;
+
+  producer.Push(TestInt{1});
+  producer.Push(TestInt{2});
+  for (int i = 0; i < 3; ++i) {
+    futures.push_back(gen());
+  }
+  ASSERT_FINISHES_OK_AND_EQ(TestInt{1}, futures[0]);
+  ASSERT_FINISHES_OK_AND_EQ(TestInt{2}, futures[1]);
+  AssertNotFinished(futures[2]);
+
+  producer.Push(TestInt{3});
+  ASSERT_FINISHES_OK_AND_EQ(TestInt{3}, futures[2]);
+  producer.Push(TestInt{4});
+  futures.push_back(gen());
+  ASSERT_FINISHES_OK_AND_EQ(TestInt{4}, futures[3]);
+  producer.Push(TestInt{5});
+  producer.Close();
+  for (int i = 0; i < 4; ++i) {
+    futures.push_back(gen());
+  }
+  ASSERT_FINISHES_OK_AND_EQ(TestInt{5}, futures[4]);
+  for (int i = 5; i < 8; ++i) {
+    ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), futures[i]);
+  }
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
+}
+
+TEST(PushGenerator, Errors) {
+  PushGenerator<TestInt> gen;
+  auto producer = gen.producer();
+  std::vector<Future<TestInt>> futures;
+
+  producer.Push(TestInt{1});
+  producer.Push(Status::Invalid("2"));
+  for (int i = 0; i < 3; ++i) {
+    futures.push_back(gen());
+  }
+  ASSERT_FINISHES_OK_AND_EQ(TestInt{1}, futures[0]);
+  ASSERT_FINISHES_AND_RAISES(Invalid, futures[1]);
+  AssertNotFinished(futures[2]);
```

Comment:
Sorry I didn't answer the earlier question (should an error always terminate a generator?).  This seems to be your test here.  I think from the general async generator concept this would be UB.  This possibility is valid.  Terminating early would also be valid.  Downstream generators should be written to expect this as a possibility and should not rely on errors terminating successive calls automatically.

Which is a long winded way of saying this is valid.

It would also be ok if `futures[2]` was `IterationTraits<TestInt>::End()` here.